File: 3953-maximum-score-with-co-prime-element/3953-maximum-score-with-co-prime-element.rb
# @param {Integer[]} nums
# @param {Integer} max_val
# @return {Integer}
def max_score(nums, max_val)
  mx = max_val

  nums.each do |x|
    mx = x if x > mx
  end

  freq = Array.new(mx + 1, 0)
  nums.each do |x|
    freq[x] += 1
  end

  spf = Array.new(mx + 1) { |i| i }
  mul = Array.new(mx + 1, 0)

  (2..mx).each do |i|
    if spf[i] == i
      j = i * 2
      while j <= mx
        spf[j] = i if spf[j] == j
        j += i
      end
    end

    cnt = 0
    j = i
    while j <= mx
      cnt += freq[j]
      j += i
    end
    mul[i] = cnt
  end

  ans = freq[1] > 0 ? 1 : 0

  mx.downto(2) do |i|
    break if i <= ans

    next if freq[i] == 0 && i > max_val

    x = i
    primes = []

    while x != 1
      p = spf[x]
      primes << p

      while (x % p).zero?
        x /= p
      end
    end

    bad = -freq[i]
    total = 1 << primes.length

    (1...total).each do |mask|
      prod = 1
      bits = 0

      primes.length.times do |k|
        if ((mask >> k) & 1) == 1
          prod *= primes[k]
          bits += 1
        end
      end

      if bits.odd?
        bad += mul[prod]
      else
        bad -= mul[prod]
      end
    end

    cost = if freq[i] > 0
             bad + freq[i] - 1
           else
             [bad, 1].max
           end

    ans = [ans, i - cost].max
  end

  ans
end

File: 3953-maximum-score-with-co-prime-element/3953-maximum-score-with-co-prime-element.swift
class Solution {
    func maxScore(_ nums: [Int], _ maxVal: Int) -> Int {
        var mx = maxVal

        for x in nums {
            if x > mx {
                mx = x
            }
        }

        var freq = Array(repeating: 0, count: mx + 1)
        for x in nums {
            freq[x] += 1
        }

        var spf = Array(0...mx)
        var mul = Array(repeating: 0, count: mx + 1)

        if mx >= 2 {
            for i in 2...mx {
                if spf[i] == i {
                    var j = i * 2
                    while j <= mx {
                        if spf[j] == j {
                            spf[j] = i
                        }
                        j += i
                    }
                }

                var cnt = 0
                var j = i
                while j <= mx {
                    cnt += freq[j]
                    j += i
                }
                mul[i] = cnt
            }
        }

        var ans = freq[1] > 0 ? 1 : 0

        var i = mx
        while i >= 2 {
            if i <= ans {
                break
            }

            if freq[i] == 0 && i > maxVal {
                i -= 1
                continue
            }

            var x = i
            var primes: [Int] = []

            while x != 1 {
                let p = spf[x]
                primes.append(p)

                while x % p == 0 {
                    x /= p
                }
            }

            var bad = -freq[i]
            let total = 1 << primes.count

            for mask in 1..<total {
                var prod = 1
                var bits = 0

                for k in 0..<primes.count {
                    if ((mask >> k) & 1) == 1 {
                        prod *= primes[k]
                        bits += 1
                    }
                }

                if (bits & 1) == 1 {
                    bad += mul[prod]
                } else {
                    bad -= mul[prod]
                }
            }

            let cost: Int
            if freq[i] > 0 {
                cost = bad + freq[i] - 1
            } else {
                cost = max(bad, 1)
            }

            ans = max(ans, i - cost)
            i -= 1
        }

        return ans
    }
}